### ISSUE TITLE
[RFC]: Modular Diffusers adapter with per-step execution via greenlet

### ISSUE THREAD

**Issue by NickCao** [vllm-project/vllm-omni#5822]
### Motivation.

DiffusersAdapterPipeline is a monolithic black box — it delegates a full generation run to one pipeline(**kwargs) call and thus does not support step-wise execution. This means: one request's entire generation blocks the engine's busy loop start to finish (no interleaving with other concurrent requests), abandoned/cancelled requests can't be stopped mid-generation, and streaming_output (progressive previews) isn't supported.

But the (not so) newly introduced Modular Diffusers changes this picture: https://huggingface.co/blog/modular-diffusers. It allows us to introspect the internals of the pipelines, and to some extent, control their executions.

We found a way to get real per-timestep control over an arbitrary diffusers Modular Pipeline's denoise loop without reimplementing it per model, validated end-to-end against QwenImage on GPU. Thus proposal to use this to implement a ModularDiffusersAdapterPipeline, with proper support for step execution.

### Proposed Change.

Proposed Change:

Add a new ModularDiffusersAdapterPipeline that loads models via diffusers' ModularPipeline framework (ModularPipeline.from_pretrained() + load_components()), covering the ~15 model families diffusers currently maps to modular pipelines (Flux, QwenImage, SDXL, Wan, SD3, etc.), opt-in via a new --diffusion-load-format modular_diffusers, and implement per-step execution.

This requires a new dependency: greenlet.

Non-goals: this does not enable cross-request GPU-batched compute (supports_request_batch/vLLM-style continuous batching). That's a hard ceiling of wrapping an opaque pipeline, not something this design fixes.

Benefit, given no batching: no head-of-line blocking across concurrent requests, responsive cancellation (stop between steps instead of burning compute on abandoned generations).

### CC List.

@alex-jw-brooks 

### Any Other Things.

PoC:
```python
#!/usr/bin/env -S uv run --script
# /// script
# requires-python = ">=3.10"
# dependencies = [
#     "torch",
#     "transformers",
#     "accelerate",
#     "greenlet",
#     "diffusers @ git+https://github.com/huggingface/diffusers",
# ]
# ///
"""POC: step-by-step execution of a diffusers ModularPipeline denoise loop
using a greenlet to pause/resume mid-`__call__`, without vLLM-Omni.

Mechanism:
  1. Resolve the pipeline's default workflow via `get_workflow()` -> a flat
     `SequentialPipelineBlocks` containing exactly one `LoopSequentialPipelineBlocks`
     (the denoise loop), per the structural guarantee `get_execution_blocks()`
     provides (it treats loop blocks as unexpandable leaves).
  2. Monkey-patch that loop block's `loop_step` (the generic, non-model-specific
     method inherited from `LoopSequentialPipelineBlocks`) so every iteration
     switches back to a "driver" greenlet after doing its real work.
  3. Run the pipeline's own unmodified `__call__` inside a "worker" greenlet.
     Diffusers drives everything -- text encoding, latent prep, the denoise
     loop's own setup/teardown, decode -- exactly as it does in the black-box
     case. The only difference is that each `loop_step()` call yields control
     back to the driver before continuing, with no re-entry, no truncated
     timesteps, and no re-running of already-completed iterations: this is a
     true pause of the whole call stack, not a restart.

Self-contained via uv (PEP 723 inline metadata) -- no manual venv setup needed:
    chmod +x poc_qwenimage_stepwise_greenlet.py
    ./poc_qwenimage_stepwise_greenlet.py --prompt "..." --steps 20

Run on a machine with the model weights reachable and a CUDA device.
"""

from __future__ import annotations

import argparse

import torch
from greenlet import greenlet

from diffusers.modular_pipelines import ComponentsManager, ModularPipeline
from diffusers.modular_pipelines.modular_pipeline import LoopSequentialPipelineBlocks


def find_loop_block(blocks) -> tuple[str, LoopSequentialPipelineBlocks]:
    """Locate the single denoise loop within a resolved workflow's blocks.

    `get_execution_blocks()`/`get_workflow()` flatten every `ConditionalPipelineBlocks`
    branch down to a concrete leaf, but explicitly stop at (never expand into)
    `LoopSequentialPipelineBlocks` -- so there is exactly one such entry here.
    """
    for name, block in blocks.sub_blocks.items():
        if isinstance(block, LoopSequentialPipelineBlocks):
            return name, block
    raise RuntimeError(f"No LoopSequentialPipelineBlocks found among: {list(blocks.sub_blocks.keys())}")


def patch_loop_step_to_yield(loop_block: LoopSequentialPipelineBlocks, driver: greenlet) -> None:
    """Make every `loop_step()` call pause the current greenlet after doing real work.

    Instance-level assignment (not a class-level def) means this does not go
    through the descriptor/auto-binding protocol, so the wrapper must accept
    the same (components, state, **kwargs) signature `loop_step` normally has.
    """
    original_loop_step = loop_block.loop_step  # bound method, inherited base-class impl

    def wrapped(components, state, **kwargs):
        components, state = original_loop_step(components, state, **kwargs)
        driver.switch(kwargs.get("i"), kwargs.get("t"), state)
        return components, state

    loop_block.loop_step = wrapped


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--model", default="Qwen/Qwen-Image")
    parser.add_argument("--prompt", default="A cat holding a sign that says hello world")
    parser.add_argument("--steps", type=int, default=20)
    parser.add_argument("--height", type=int, default=512)
    parser.add_argument("--width", type=int, default=512)
    parser.add_argument("--seed", type=int, default=0)
    parser.add_argument("--device", default="cuda")
    parser.add_argument("--out", default="qwenimage_stepwise_poc.png")
    args = parser.parse_args()

    print(f"Loading {args.model} as a ModularPipeline...")
    manager = ComponentsManager()
    pipe = ModularPipeline.from_pretrained(args.model, components_manager=manager)
    pipe.load_components(torch_dtype=torch.bfloat16)
    pipe.to(args.device)

    resolved = pipe.blocks.get_workflow("text2image")
    print(f"Resolved 'text2image' workflow stages: {list(resolved.sub_blocks.keys())}")

    loop_name, loop_block = find_loop_block(resolved)
    print(f"Denoise loop stage: '{loop_name}' ({type(loop_block).__name__})")

    driver = greenlet.getcurrent()
    patch_loop_step_to_yield(loop_block, driver)

    # Swap in the resolved, single-workflow block graph. `pipe.__call__` seeds
    # `PipelineState` from `self._blocks.inputs` and calls `self._blocks(self, state)`,
    # so from here on diffusers drives the whole thing unmodified.
    pipe._blocks = resolved

    call_kwargs = dict(
        prompt=args.prompt,
        num_inference_steps=args.steps,
        height=args.height,
        width=args.width,
        generator=torch.Generator(device=args.device).manual_seed(args.seed),
        output_type="pil",
    )

    def run_pipeline():
        return pipe(**call_kwargs)

    worker = greenlet(run_pipeline)

    step_count = 0
    result = worker.switch()  # starts run_pipeline(); runs until the first loop_step yield
    while not worker.dead:
        i, t, _block_state = result
        step_count += 1
        print(f"[driver] intercepted step {step_count} (i={i}, t={float(t):.3f}) -- control back in our hands here")
        # This is exactly where vLLM-Omni's engine would return to its busy
        # loop: check for aborts, admit other requests, stream a partial
        # preview, swap encoder/DiT offload state, etc.
        result = worker.switch()  # resume the worker for the next iteration

    output = result  # worker finished normally; this is pipe(**call_kwargs)'s return value
    print(f"Done: {step_count} steps intercepted via greenlet, then decode ran automatically.")
    output.images[0].save(args.out)
    print(f"Saved {args.out}")


if __name__ == "__main__":
    main()
```

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://vllm-omni.readthedocs.io), which can answer lots of frequently asked questions.

**Comment by alex-jw-brooks** [comment]
I think this is very interesting and could be really useful. I'm not familiar with greenlet, but do you anticipate any concurrency issues with integrating this alongside the asyncio event loop?

Also, I think it would be useful to know the performance of batch_size=1 performance of this PoC against the current integration to make sure it doesn't hurt performance in the degenerate case. It looks like greenlet is pretty lightweight, but would be nice to verify that there isn't unexpected copy / serialization costs

CC @fhfuih @hsliuustc0106 @sayakpaul

**Comment by NickCao** [comment]
> I think this is very interesting and could be really useful. I'm not familiar with greenlet, but do you anticipate any concurrency issues with integrating this alongside the asyncio event loop?

Whatever concurrent execution mechanism we use, we do need to deepcopy the stateful components (e.g. the scheduler) to prevent cross talk between requests.

> Also, I think it would be useful to know the performance of batch_size=1 performance of this PoC against the current integration to make sure it doesn't hurt performance in the degenerate case.

The performance hit is expected to be minimal, just some additional context (stack) switching.

> It looks like greenlet is pretty lightweight, but would be nice to verify that there isn't unexpected copy / serialization costs

As said in the first point, we do have to copy some (lightweight) components, benchmark will tell.

**Comment by NickCao** [comment]
See https://github.com/NickCao/vllm-omni/tree/5822-poc-v1 for a full PoC (batching TBD):

```
vllm serve --omni Qwen/Qwen-Image --diffusion-load-format modular_diffusers --step-execution --diffusion-attention-backend TORCH_SDPA
``` 

And try to send and abort requests, see GPU usage instantly drop.

**Comment by fhfuih** [comment]
Thanks for the RFC! Yes, when I was implementing the initial diffusers backend, @sayakpaul also mentioned this feature (#2403). If I remember it correctly, the modular feature is super new if not experimental at that time, so I didn't bother using it. But it's nice if you can pick it up.

I think you don't need to add yet another user-facing `modular_diffusers` toggle. Just override the current diffusers implementation (both are diffusers after all!). Or if you want to keep the current simpler/naive implementation for any reason (i.e. breaking change to the new modular diffusion pipeline I guess?), you can assume `--diffusion-load-format diffusers` + `--step-execution` = your new impl.

For the greenlet part, so it is primarily to intercept per-step output and pause at it (until we call `worker.switch()`, after which it pauses at next interception), right? I wonder if there is any simpler way (e.g., does the modular pipeline itself offer any interface to run only one loop for the loop component)

Just FYI, modular pipeline effectively offers the capability to check input-output modality combinations available from the pipeline. In my previous RFC, diffusers contributors also suggested making use of it if we want more user-friendly checking.

**Comment by NickCao** [comment]
> Or if you want to keep the current simpler/naive implementation for any reason

Mostly due to the modular pipeline's experimental status (as of writing still documented as such).

> For the greenlet part, so it is primarily to intercept per-step output and pause at it (until we call worker.switch(), after which it pauses at next interception), right?

Yes

> I wonder if there is any simpler way (e.g., does the modular pipeline itself offer any interface to run only one loop for the loop component)

Not really, the loop handling is internal to LoopSequentialPipelineBlocks.

> Just FYI, modular pipeline effectively offers the capability to check input-output modality combinations available from the pipeline. In my previous RFC, diffusers contributors also suggested making use of it if we want more user-friendly checking.

Nice! Maybe we can start with an implementation that just makes use of that.

**Comment by sayakpaul** [comment]
I am no greenlet expert. Just know what it does. I would appreciate why specifically we need `greenlet` here.

It's great to see modular diffusers picking up traction :) Modular has been graduated out of the experimental stint, btw. All the new important and impactful releases now use it: H3, LTX-2.5, Wan Animate 2, and others that are upcoming. So, you can consider it robust.

We have beefed up our testing suite to make sure it's robust enough.

We will be more than happy to help with this.

**Comment by yiyixuxu** [comment]
hi!

a heads-up from diffusres side, we have this PR upcoming and i think it's relevant https://github.com/huggingface/diffusers/pull/14159

I just added first-class streaming support to the PR after I saw this RFC (it's been on my TODO for a while:) 

will be ready for review soon, onnce it is, i'd love to hear your feedbacks
